### DISCUSSION TITLE
Match case command in SimpleSerialShell

### DISCUSSION THREAD

**Discussion by lc2047** [philj404/SimpleSerialShell#31] [Q&A] [answered]
Hi

I tried SimpleSerialShell, but the shell doesn't care about the case of the command(input string).
What do I do to have case sensitive(match case) function?

Please share your comments with me.

Sincerely

LC

**Comment by philj404** [comment, marked as answer]
**What you would need to change:**  

See [SimpleSerialShell.cpp](https://github.com/philj404/SimpleSerialShell/blob/main/src/SimpleSerialShell.cpp), the `compareName()` method at about line 59.

The `strncasecmp(...)` ignores case while comparing the entered text against the list of commands.  If you replace it with `strncmp(...)` commands will not ignore case -- "run", "Run" and "RUN" will each be unique commands.

I use `strncasecmp()` in the library because I feel it is important that an emergency command such as `stop` should execute, even if I get excited and I bump the caps lock key.  Any of `stop`, `STOP` and even `stOP` should make the machine stop!

**How to make the change:**  

My requirements may not make sense for your application.  There are several ways to customize this library for your particular needs.

- The command line arguments are passed into the command.  For a particular command that needs to care about case, the command could figure out which sub-command to run by reading `argv[0]` itself.
- If case comparison is ALWAYS important for ALL commands, you can just copy the two source files into your sketch and modify them as needed.  (Note if you do this your code will not pick up library updates... but the code is pretty stable now.)
- If you have several sketches that need this change, you can create your own fork of the library and make your changes there.

I hope this helps.

**Reply by lc2047** [reply]
Hi philj404

You are very nice and thanks for your help. 
sorry, I am a beginner in programming and my English is not good. So not clear in reading. Usually, some experts just require me to reading more and everything has been described in a document/textbook. That is true and I tried, but I am still not clear. So ask them for help.

To return for your comments, I agree with your consideration. But my application, the shell is provided for humans and another program. Therefore the case-sensitive issue is accepted.

Sincerely

LC
